### PR TITLE
Wire MoE and MXFP8Linear training paths to CuTeDSL by default

### PR DESCRIPTION
**Pull request by alexsamardzic**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #4620
* #4604
* #4594
* #4589
* #4585

### CODE REVIEW

**Comment by alexsamardzic** [comment]
This PR wires the MXFP8 MoE grouped-mm and MXFP8Linear training paths to use the CuTeDSL backend by default, while keeping the legacy CUDA/Triton backend selectable for validation and benchmarking. It adds end-to-end correctness tests for grouped-mm, padded grouped-mm, prequantized MXTensor grouped-mm, and MXFP8Linear fwd/bwd, plus benchmark support for comparing CuTeDSL against the legacy backend. This prepares the next PR to delete torchao._C_mxfp8 and the now-obsolete legacy CUDA/Triton paths.

---

To run tests:
```sh
pytest -q \
    test/prototype/moe_training/test_mxfp8_grouped_mm.py::test_mxfp8_grouped_gemm_backend_fwd_bwd \
    test/prototype/moe_training/test_mxfp8_grouped_mm.py::test_mxfp8_grouped_gemm_padded_backend_fwd_bwd \
    test/prototype/moe_training/test_mxfp8_grouped_mm.py::test_mxfp8_grouped_gemm_mxtensor_backend_forward \
    test/prototype/moe_training/test_mxfp8_grouped_mm.py::test_mxfp8_grouped_gemm_with_dq_fwd_bwd \
    test/prototype/moe_training/test_mxfp8_linear.py::test_mxfp8_linear_fwd_bwd_sqnr
```

---

To run MoE grouped-mm fwd/bwd recipe benchmark:
```sh
python benchmarks/prototype/moe_training/mxfp8/roofline_unified.py \
    --backend=both --K=4096 --N=4096 --G=8 --breakdown_M=16384 \
    --outfile_speedup=roofline_speedup_K4096_N4096.csv \
    --outfile_quant_2d=roofline_quant_2d_K4096_N4096.csv \
    --outfile_quant_3d=roofline_quant_3d_K4096_N4096.csv \
    --plot_file=roofline_K4096_N4096.png
```

Results of benchmark above:
roofline_K4096_N4096

---

To run MXFP8Linear fwd/bwd recipe benchmark:
```sh
python scripts/validate_mxfp8_linear_cutedsl.py \
    --M 4096 --N 4096 --K 2048 \
    --cuda-graph-bench --graph-iters 100
```

---

Would be good to run on multi-GPU machine:
```sh
torchrun --nproc-per-node=2 --local-ranks-filter=0 \
    benchmarks/prototype/moe_training/mxfp8/bench_ep_pipeline.py --backend=legacy

torchrun --nproc-per-node=2 --local-ranks-filter=0 \
    benchmarks/prototype/moe_training/mxfp8/bench_ep_pipeline.py --backend=cutedsl
```

**Comment by alexsamardzic** [comment]
Added `--calibrate_roofline` option to `roofline_unified.py`. When turned on, the script measures achievable copy bandwidth plus BF16/FP8 GEMM throughput on the current GPU and uses those calibrated values for the roofline model, instead of relying only on the hard-coded B200 assumptions. The output and plot labels now refer to these values as achievable roofline ceilings.

**Comment by andrewor14** [comment]
Hi @alexsamardzic, is this the PR that's incorrectly wired? Should we close it then?

**Comment by alexsamardzic** [comment]
Replaced by #4705.

**Comment by alexsamardzic** [comment]
> Hi @alexsamardzic, is this the PR that's incorrectly wired? Should we close it then?

All five on the stack were wrong, I used plain `ghstack` command to create them.  I had to close all of them, and replaced them with another stack, created through `ghstack --direct`, #4705 is replacement for this one, and the rest is in the same new stack.  I apologize for the mess - previously I was creating proper branches for my PRs here, and in this case I used ghstakh out of habit from mostly working with PyTorch repo.

Btw: All five of these PRs are reasonably simple, I've stated test plans, and also benchmarking results showing that the perfomance is unchanged, or better - so I hope we could merge them quickly.  The top one, #4705, is the only one to be careful about, as this one switches the default to new CuTeDSL kernels.  After we merge this, and then if we see for couple weeks that there are no issues, I'll post another PR to delete legacy kernels and the C++ extension.
